File: 10-code-control/16-percentage-rollout/python-console/16-percentage-rollout.py
#!/usr/bin/env python3
"""Console grid navigator for a static LaunchDarkly percentage rollout."""

from __future__ import annotations

import curses
import os
import sys
from pathlib import Path
from typing import Any

import ldclient
from ldclient import Config
from ldclient.client import LDClient

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
from rollout import evaluate_rollout  # noqa: E402

# LaunchDarkly: percentage rollout — sticky assignment by context key.
# https://launchdarkly.com/docs/home/flags/rollouts

ROWS = ("t", "m", "b")
COLS = ("l", "m", "r")
APP_BANNER = "16-percentage-rollout[python-console]"
CONFIGURED_GREEN_PCT = 30
HISTORY_LIMIT = 8
GREEN_PAIR = 5
INK_PAIR = 7

_ld_client: LDClient | None = None


def init_launchdarkly() -> None:
    """Initialize one process-wide SDK client for percentage evaluation."""
    global _ld_client
    sdk_key = (os.environ.get("LD_SDK_KEY") or "").strip()
    if not sdk_key:
        return
    ldclient.set_config(Config(sdk_key))
    _ld_client = ldclient.get()


def generated_username(base: str, index: int) -> str:
    """Map a sequence index onto a context key. 0 is the original login."""
    if index <= 0:
        return base
    return f"{base}{index}"


def format_pos(row: int, col: int) -> str:
    return f"{ROWS[row]}/{COLS[col]}"


def try_move(row: int, col: int, dr: int, dc: int) -> tuple[int, int, bool]:
    new_row = max(0, min(2, row + dr))
    new_col = max(0, min(2, col + dc))
    return new_row, new_col, (new_row, new_col) != (row, col)


def put(
    stdscr: curses.window, y: int, x: int, text: str, attr: int = curses.A_NORMAL
) -> None:
    try:
        stdscr.addstr(y, x, text, attr)
    except curses.error:
        pass


def cell_attr(color: str) -> int:
    if color == "green":
        return curses.color_pair(GREEN_PAIR)
    return curses.A_NORMAL


def read_username(stdscr: curses.window) -> str:
    stdscr.timeout(-1)
    curses.echo()
    stdscr.clear()
    put(stdscr, 0, 0, APP_BANNER)
    put(stdscr, 2, 0, "Login")
    put(stdscr, 3, 0, "Username becomes the LaunchDarkly user context key.")
    put(stdscr, 5, 0, "Username: ")
    stdscr.refresh()
    while True:
        username = stdscr.getstr(5, 10, 40).decode("utf-8").strip()
        if username:
            curses.noecho()
            return username
        put(stdscr, 7, 0, "Username is required. Try again.")
        stdscr.clrtoeol()
        stdscr.refresh()


def draw_cell(
    stdscr: curses.window, y: int, x: int, selected: bool, color: str
) -> None:
    top = "┏━━━┓" if selected else "┌───┐"
    mid = "┃ X ┃" if selected else "│   │"
    bot = "┗━━━┛" if selected else "└───┘"
    attr = cell_attr(color) if selected else curses.A_NORMAL
    put(stdscr, y, x, top, attr)
    put(stdscr, y + 1, x, mid, attr)
    put(stdscr, y + 2, x, bot, attr)


def observed_line(seen: dict[str, str]) -> str:
    total = len(seen)
    if not total:
        return "Observed: no unique keys yet."
    green_count = sum(1 for value in seen.values() if value == "green")
    none_count = total - green_count
    pct = round((green_count / total) * 1000) / 10
    return (
        f"Observed: {green_count} green / {none_count} none of {total} unique "
        f"→ {pct}% (configured {CONFIGURED_GREEN_PCT}%)"
    )


def nav_hint(next_name: str, previous_name: str | None) -> str:
    prev = "—" if previous_name is None else previous_name
    return (
        f"Arrows/WASD move.  N next ({next_name})  P prev ({prev})  "
        "L logout  Q quit"
    )


def draw_screen(
    stdscr: curses.window,
    username: str,
    next_name: str,
    previous_name: str | None,
    row: int, col: int,
    previous: tuple[int, int] | None,
    flags: dict[str, Any],
    seen: dict[str, str],
    history: list[dict[str, str]],
) -> None:
    stdscr.clear()
    stdscr.bkgd(" ", curses.color_pair(INK_PAIR))
    color = str(flags.get("highlightColor") or "none")
    reason = flags.get("reason") or {}
    reason_kind = str(reason.get("kind") or "UNKNOWN")
    variation = flags.get("variationIndex")
    variation_text = "default" if variation is None else str(variation)

    put(stdscr, 0, 0, APP_BANNER)
    put(stdscr, 1, 0, "Name: ")
    put(stdscr, 1, 6, username, cell_attr(color))
    prev_text = format_pos(*previous) if previous else "—"
    put(
        stdscr,
        2,
        0,
        f"Current: {format_pos(row, col)}   Previous: {prev_text}",
    )
    put(
        stdscr,
        3,
        0,
        f"Flag value: {flags.get('flagValue')}   reason: {reason_kind}   "
        f"idx: {variation_text}",
    )
    put(stdscr, 4, 0, observed_line(seen))
    put(stdscr, 5, 0, nav_hint(next_name, previous_name))

    base_y, base_x = 7, 2
    for r in range(3):
        for c in range(3):
            selected = r == row and c == col
            draw_cell(
                stdscr,
                base_y + r * 4,
                base_x + c * 6,
                selected,
                color if selected else "none",
            )

    hist_x = 24
    put(stdscr, base_y, hist_x, "Recent evaluations")
    if not history:
        put(stdscr, base_y + 1, hist_x, "(none yet)", curses.A_DIM)
    for index, item in enumerate(history):
        label = f"{item['username']}  {item['value']}  {item['reason']}"
        attr = cell_attr(item["value"])
        put(stdscr, base_y + 1 + index, hist_x, label, attr)
    stdscr.refresh()


def remember(
    username: str,
    flags: dict[str, Any],
    seen: dict[str, str],
    history: list[dict[str, str]],
) -> None:
    value = str(flags.get("flagValue") or "none")
    reason = str((flags.get("reason") or {}).get("kind") or "UNKNOWN")
    seen[username] = value
    history[:] = [item for item in history if item["username"] != username]
    history.insert(0, {"username": username, "value": value, "reason": reason})
    del history[HISTORY_LIMIT:]


def run_grid(stdscr: curses.window, base_username: str) -> str:
    """Navigate the grid and walk generated usernames with N / P.

    Index 0 is the original login. N steps forward (marco → marco1 → marco2);
    P steps back and stops at the login name. LaunchDarkly still owns
    assignment; this loop only changes the context key and re-evaluates.
    """
    index = 0
    username = generated_username(base_username, index)
    row, col = 1, 1
    previous: tuple[int, int] | None = None
    seen: dict[str, str] = {}
    history: list[dict[str, str]] = []
    stdscr.timeout(500)
    try:
        while True:
            flags = evaluate_rollout(_ld_client, username)
            remember(username, flags, seen, history)
            draw_screen(
                stdscr,
                username,
                generated_username(base_username, index + 1),
                None if index == 0 else generated_username(base_username, index - 1),
                row,
                col,
                previous,
                flags,
                seen,
                history,
            )
            key = stdscr.getch()
            if key == -1:
                continue
            if key in (ord("q"), ord("Q")):
                return "quit"
            if key in (ord("l"), ord("L")):
                return "logout"
            if key in (ord("n"), ord("N")):
                index += 1
                username = generated_username(base_username, index)
                row, col = 1, 1
                previous = None
                continue
            if key in (ord("p"), ord("P")):
                if index == 0:
                    continue
                index -= 1
                username = generated_username(base_username, index)
                row, col = 1, 1
                previous = None
                continue

            dr = dc = 0
            if key in (curses.KEY_UP, ord("w"), ord("W")):
                dr = -1
            elif key in (curses.KEY_DOWN, ord("s"), ord("S")):
                dr = 1
            elif key in (curses.KEY_LEFT, ord("a"), ord("A")):
                dc = -1
            elif key in (curses.KEY_RIGHT, ord("d"), ord("D")):
                dc = 1
            else:
                continue

            new_row, new_col, moved = try_move(row, col, dr, dc)
            if moved:
                previous = (row, col)
                row, col = new_row, new_col
    finally:
        stdscr.timeout(-1)


def init_colors() -> None:
    curses.start_color()
    curses.use_default_colors()
    curses.init_pair(GREEN_PAIR, curses.COLOR_GREEN, curses.COLOR_BLACK)
    curses.init_pair(INK_PAIR, curses.COLOR_WHITE, curses.COLOR_BLACK)


def main(stdscr: curses.window) -> None:
    curses.curs_set(0)
    init_colors()
    stdscr.keypad(True)
    init_launchdarkly()
    try:
        while True:
            username = read_username(stdscr)
            if run_grid(stdscr, username) == "quit":
                break
    finally:
        if _ld_client is not None:
            _ld_client.close()


if __name__ == "__main__":
    curses.wrapper(main)

File: 10-code-control/16-percentage-rollout/node-console/16-percentage-rollout.js
#!/usr/bin/env node
/** Console grid navigator for a static LaunchDarkly percentage rollout. */

const readline = require("readline");
const LaunchDarkly = require("@launchdarkly/node-server-sdk");
const { evaluateRollout } = require("../rollout");

// LaunchDarkly: percentage rollout — sticky assignment by context key.
// https://launchdarkly.com/docs/home/flags/rollouts

const ROWS = ["t", "m", "b"];
const COLS = ["l", "m", "r"];
const APP_BANNER = "16-percentage-rollout[node-console]";
const CONFIGURED_GREEN_PCT = 30;
const HISTORY_LIMIT = 8;
const GREEN = "\x1b[92m";
const BG = "\x1b[48;5;236m";
const RESET = "\x1b[0m";

let ldClient = null;

async function initLaunchDarkly() {
  const sdkKey = (process.env.LD_SDK_KEY || "").trim();
  if (!sdkKey) return;
  ldClient = LaunchDarkly.init(sdkKey);
  try {
    await ldClient.waitForInitialization({ timeout: 5 });
  } catch (_) {
    ldClient.close();
    ldClient = null;
  }
}

function generatedUsername(base, index) {
  return index <= 0 ? base : `${base}${index}`;
}

function formatPos(row, col) {
  return `${ROWS[row]}/${COLS[col]}`;
}

function tryMove(row, col, dr, dc) {
  const nr = Math.max(0, Math.min(2, row + dr));
  const nc = Math.max(0, Math.min(2, col + dc));
  return { row: nr, col: nc, moved: nr !== row || nc !== col };
}

function colorize(text, color) {
  if (color !== "green") return text;
  return `${GREEN}${text}${RESET}${BG}`;
}

function drawCell(selected, color) {
  if (selected) {
    const lines = ["┏━━━┓", "┃ X ┃", "┗━━━┛"];
    return color === "green" ? lines.map((line) => colorize(line, color)) : lines;
  }
  return ["┌───┐", "│   │", "└───┘"];
}

function observedLine(seen) {
  const total = seen.size;
  if (!total) return "Observed: no unique keys yet.";
  let greenCount = 0;
  for (const value of seen.values()) if (value === "green") greenCount += 1;
  const noneCount = total - greenCount;
  const pct = Math.round((greenCount / total) * 1000) / 10;
  return (
    `Observed: ${greenCount} green / ${noneCount} none of ${total} unique ` +
    `→ ${pct}% (configured ${CONFIGURED_GREEN_PCT}%)`
  );
}

function navHint(nextName, previousName) {
  const prev = previousName || "—";
  return `Arrows/WASD move.  N next (${nextName})  P prev (${prev})  L logout  Q quit`;
}

function remember(username, flags, seen, history) {
  const value = flags.flagValue || "none";
  const reason = (flags.reason && flags.reason.kind) || "UNKNOWN";
  seen.set(username, value);
  const next = history.filter((item) => item.username !== username);
  next.unshift({ username, value, reason });
  history.length = 0;
  history.push(...next.slice(0, HISTORY_LIMIT));
}

function render(state, flags) {
  const { username, index, base, row, col, previous, seen, history } = state;
  process.stdout.write(`${BG}\x1b[2J\x1b[H`);
  const color = flags.highlightColor || "none";
  const reasonKind = (flags.reason && flags.reason.kind) || "UNKNOWN";
  const variation = flags.variationIndex == null ? "default" : String(flags.variationIndex);
  const prevText = previous ? formatPos(previous.row, previous.col) : "—";
  console.log(APP_BANNER);
  console.log(`Name: ${colorize(username, color)}${RESET}${BG}`);
  console.log(`Current: ${formatPos(row, col)}   Previous: ${prevText}`);
  console.log(`Flag value: ${flags.flagValue}   reason: ${reasonKind}   idx: ${variation}`);
  console.log(observedLine(seen));
  console.log(
    navHint(
      generatedUsername(base, index + 1),
      index === 0 ? null : generatedUsername(base, index - 1)
    )
  );
  console.log("");
  for (let r = 0; r < 3; r++) {
    const top = [];
    const mid = [];
    const bot = [];
    for (let c = 0; c < 3; c++) {
      const selected = r === row && c === col;
      const lines = drawCell(selected, selected ? color : "none");
      top.push(lines[0]);
      mid.push(lines[1]);
      bot.push(lines[2]);
    }
    console.log(top.join(" "));
    console.log(mid.join(" "));
    console.log(bot.join(" "));
  }
  console.log("");
  console.log("Recent evaluations");
  if (!history.length) console.log("(none yet)");
  for (const item of history) {
    console.log(colorize(`${item.username}  ${item.value}  ${item.reason}`, item.value));
  }
}

function askUsername() {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  return new Promise((resolve) => {
    const prompt = () => {
      rl.question("Username: ", (answer) => {
        const name = answer.trim();
        if (!name) {
          console.log("Username is required.");
          prompt();
          return;
        }
        rl.close();
        resolve(name);
      });
    };
    console.log(APP_BANNER);
    console.log("Login");
    console.log("Username becomes the LaunchDarkly user context key.\n");
    prompt();
  });
}

function runGrid(base) {
  const state = {
    base,
    index: 0,
    username: generatedUsername(base, 0),
    row: 1,
    col: 1,
    previous: null,
    seen: new Map(),
    history: [],
  };

  return new Promise((resolve) => {
    let active = true;

    const onKeypress = async (str, key) => {
      if (!active) return;
      if (key.ctrl && key.name === "c") {
        cleanup();
        resolve("quit");
        return;
      }
      if (str === "q" || str === "Q") {
        cleanup();
        resolve("quit");
        return;
      }
      if (str === "l" || str === "L") {
        cleanup();
        resolve("logout");
        return;
      }
      if (str === "n" || str === "N") {
        state.index += 1;
        state.username = generatedUsername(base, state.index);
        state.row = 1;
        state.col = 1;
        state.previous = null;
        await refresh();
        return;
      }
      if (str === "p" || str === "P") {
        if (state.index === 0) return;
        state.index -= 1;
        state.username = generatedUsername(base, state.index);
        state.row = 1;
        state.col = 1;
        state.previous = null;
        await refresh();
        return;
      }

      let dr = 0;
      let dc = 0;
      if (key.name === "up" || str === "w" || str === "W") dr = -1;
      else if (key.name === "down" || str === "s" || str === "S") dr = 1;
      else if (key.name === "left" || str === "a" || str === "A") dc = -1;
      else if (key.name === "right" || str === "d" || str === "D") dc = 1;
      else return;

      const result = tryMove(state.row, state.col, dr, dc);
      if (result.moved) {
        state.previous = { row: state.row, col: state.col };
        state.row = result.row;
        state.col = result.col;
      }
      await refresh();
    };

    let pollTimer = null;

    function cleanup() {
      active = false;
      if (pollTimer) clearInterval(pollTimer);
      pollTimer = null;
      process.stdin.removeListener("keypress", onKeypress);
      if (process.stdin.isTTY) process.stdin.setRawMode(false);
      process.stdin.pause();
    }

    async function refresh() {
      if (!active) return;
      const flags = await evaluateRollout(ldClient, state.username);
      if (!active) return;
      remember(state.username, flags, state.seen, state.history);
      render(state, flags);
    }

    readline.emitKeypressEvents(process.stdin);
    if (process.stdin.isTTY) process.stdin.setRawMode(true);
    process.stdin.resume();
    process.stdin.on("keypress", onKeypress);
    refresh().then(() => {
      pollTimer = setInterval(refresh, 500);
    });
  });
}

async function main() {
  await initLaunchDarkly();
  while (true) {
    const username = await askUsername();
    const action = await runGrid(username);
    if (action === "quit") {
      if (ldClient) ldClient.close();
      process.exit(0);
    }
  }
}

main();
